File: Backend/project/apps/financings/class/credit.py
from datetime import datetime, timedelta
from dateutil.relativedelta import relativedelta

class Credit:
    def __init__(self, proposito, monto, plazo, tasa_interes, forma_de_pago, frecuencia_pago, fecha_inicio, tipo_credito, destino_id, customer_id, fecha_vencimiento=None):
        self.__proposito = proposito
        self.__monto = monto
        self.__plazo = plazo
        self.__tasa_interes = tasa_interes
        self.__forma_de_pago = forma_de_pago
        self.__frecuencia_pago = frecuencia_pago
        self.__fecha_inicio = datetime.strptime(fecha_inicio, '%Y-%m-%d')
        self.__fecha_vencimiento = fecha_vencimiento if fecha_vencimiento else self.calcular_fecha_vencimiento()
        self.__tipo_credito = tipo_credito
        self.__destino_id = destino_id
        self.__customer_id = customer_id

    @property
    def frecuencia_pago(self):
        return self.__frecuencia_pago
    
    @frecuencia_pago.setter
    def frecuencia_pago(self, value):
        frecuenciaPago = ['MENSUAL', 'TRIMESTRAL', 'SEMANAL']
        if value in frecuenciaPago:
            self.__frecuencia_pago = value
        else:
            print("Frecuencia de pago no válida")

    @property
    def plazo(self):
        return self.__plazo
    
    @plazo.setter
    def plazo(self, value):
        self.__plazo = value
    
    @property
    def fecha_inicio(self):
        return self.__fecha_inicio
    
    @fecha_inicio.setter
    def fecha_inicio(self, value):
        formato = '%Y-%m-%d'
        try:
            self.__fecha_inicio = datetime.strptime(value, formato)
        except ValueError:
            self.__fecha_inicio = datetime.now()

    def calcular_fecha_vencimiento(self):
        fecha_inicio = self.__fecha_inicio
        plazo = self.__plazo
        fecha_vencimiento = fecha_inicio + relativedelta(months=plazo)
        return fecha_vencimiento.strftime('%Y-%m-%d')

File: Backend/project/static/js/financings/credit/credit.js
import { Credit } from '../../class/credit.js';
import { PaymentPlan } from '../../class/paymentplan.js';

const proposito = document.getElementById('proposito');
const monto = document.getElementById('monto');
const plazo = document.getElementById('plazo');
const tasa_interes = document.getElementById('tasa_interes');
const forma_de_pago = document.getElementById('forma_de_pago');
// const frecuencia_pago = document.getElementById('frecuencia_pago');

const fecha_inicio = document.getElementById('fecha_inicio');
const tipo_credito = document.getElementById('tipo_credito');
const customer_id = document.getElementById('customer_id');
const tbody_plan = document.getElementById('tbody_plan');

function generar_plan() {
    tbody_plan.innerHTML = '';
  
    const fechaInicioValue = new Date(fecha_inicio.value);
    

    let credito = new Credit(proposito.value, monto.value, plazo.value, tasa_interes.value, forma_de_pago.value, 'MENSUAL', fechaInicioValue, tipo_credito.value, null, customer_id.value);
    
    let plan_pago = new PaymentPlan(credito);
    let plan = plan_pago.generarPlan();
    console.log(credito.toJSON());
    plan.forEach(element => {
        var nueva_fila = tbody_plan.insertRow();
        var mes = nueva_fila.insertCell(0);
        mes.textContent = element['mes'];
        var fechaIni = nueva_fila.insertCell(1);
        fechaIni.textContent = transformarFecha(element['fecha inicio']);
        var fechaVenc = nueva_fila.insertCell(2);
        fechaVenc.textContent = transformarFecha(element['fecha final']);
        var monto = nueva_fila.insertCell(3);
        monto.textContent = 'Q' + element['monto_prestado'];
        var interes = nueva_fila.insertCell(4);
        interes.textContent = 'Q' + element['intereses'];
        var capital = nueva_fila.insertCell(5);
        capital.textContent = 'Q' + element['capital'];
        var cuota = nueva_fila.insertCell(6);
        cuota.textContent = 'Q' + element['cuota'];
    });
}

function transformarFecha(ele) {
    const fecha = new Date(ele);
    const dia = fecha.getDate().toString().padStart(2, '0');
    const mes = (fecha.getMonth() + 1).toString().padStart(2, '0');
    const año = fecha.getFullYear();
    return `${dia}/${mes}/${año}`;
}

document.getElementById('generar_plan').onclick = generar_plan;

document.getElementById('credito').addEventListener('submit', async function(event) {
    event.preventDefault();
    try {
        let credito = new Credit();
        credito.proposito = proposito.value;
        credito.monto = monto.value;
        credito.plazo = plazo.value;
        credito.tasaInteres = tasa_interes.value; // Asegúrate de que este valor es el correcto
        credito.formaDePago = forma_de_pago.value;
        credito.frecuenciaPago = 'MENSUAL';
        credito.fechaInicio = new Date(fecha_inicio.value);
        credito.tipoCredito = tipo_credito.value;
        credito.customerId = customer_id.value;
        credito.fechaVencimiento = new Date(fecha_inicio.value); // Corregido para obtener la fecha de vencimiento correcta
        credito.fechaVencimiento.setFullYear(credito.fechaVencimiento.getFullYear() + 1);
        
        console.log(credito.toJSON());

        const credi = await registrarCredito('http://127.0.0.1:8000/financings/api/credit/', credito);
        console.log('Credito Registrado', credi);
        alert('¡Formulario enviado con éxito!');
        window.location.href = '/customers/';
    } catch (error) {
        console.error('Error al registrar los datos:', error);
        alert('Hubo un error al enviar el formulario. Por favor, inténtalo de nuevo.');
    }
});

async function registrarCredito(url, credito) {
    try {
        let json = {
            customer_id: credito.customerId,
            destino_id: credito.destinoId,
            fecha_inicio: credito.fechaInicio.toISOString().split('T')[0],
            fecha_vencimiento: credito.fechaVencimiento.toISOString().split('T')[0],
            forma_de_pago: credito.formaDePago,
            frecuencia_pago: credito.frecuenciaPago,
            monto: credito.monto,
            plazo: credito.plazo,
            proposito: credito.proposito,
            tasa_interes: credito.tasaInteres, // Asegúrate de que este valor es el correcto
            tipo_credito: credito.tipoCredito
        };
        console.log(json);

        // Obtener el token CSRF del meta tag
        const csrfTokenElement = document.querySelector('meta[name="csrf-token"]');
        if (!csrfTokenElement) {
            throw new Error('CSRF token not found');
        }
        const csrfToken = csrfTokenElement.getAttribute('content');

        const response = await fetch(url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'X-CSRFToken': csrfToken // Incluir el token CSRF en las cabeceras
            },
            body: JSON.stringify(json)
        });

        if (!response.ok) {
            throw new Error(`Error: ${response.status}`);
        }

        const data = await response.json();
        console.log(data);
        return data;
    } catch (error) {
        console.error('Error:', error);
        throw error;
    }
}
